Adds Decompositions of Common Gates

Lucaman99 opened the pull request:
**Context:**

https://github.com/XanaduAI/pennylane/issues/644

**Description of the Change:**

Adds `decomposition` method to PauliX, PauliY, PauliZ, S, T, and Hadamard gates, which decomposes each of these gates into a product of rotation gates.

**Benefits:**

Allows gate decompositions to take place within the PennyLane core, which is helpful when dealing with multiple plugins with multiple different gate sets.

**Possible Drawbacks:**

None.

**Related GitHub Issues:**

https://github.com/XanaduAI/pennylane/issues/644

josh146 commented on pennylane/ops/qubit.py:
```diff
@@ -73,6 +73,11 @@ def diagonalizing_gates(self):
 
+    @staticmethod
+    def decomposition(wires):
+        decomp_ops = [RX(np.pi / 2, wires=wires), RZ(np.pi / 2, wires=wires)]
```
Unless I'm looking at it incorrectly, this doesn't seem to work for me 🤔 
```python
>>> qml.RZ(np.pi/2, wires=0).matrix @ qml.RX(np.pi/2, wires=0).matrix
array([[ 0.5-0.5j, -0.5-0.5j],
       [ 0.5-0.5j,  0.5+0.5j]])
```

However, a 180 degree Y rotation followed by a 90 degree X rotation works:

```python
>>> qml.PauliX(0).matrix @ qml.RY(np.pi/2, wires=0).matrix
array([[ 0.70710678,  0.70710678],
       [ 0.70710678, -0.70710678]])
```

Lucaman99 replied:
Yes, you're right, I meant to do Rz(pi/2) Rx(pi/2) Rz(pi/2), but the one you provide in more simple 🙂

josh146 commented on pennylane/ops/qubit.py:
```diff
@@ -223,6 +243,11 @@ def _matrix(cls, *params):
 
+    @staticmethod
+    def decomposition(wires):
+        decomp_ops = [RZ(np.pi / 2, wires=wires)]
```
Note that by decomposing this in terms of the phase shift gate, we avoid the global phase:

```python
>>> np.round(qml.PhaseShift(np.pi / 2, wires=0).matrix, 3)
array([[1.+0.j, 0.+0.j],
       [0.+0.j, 0.+1.j]])
```

I'm wondering if it makes sense to decompose this in terms of `qml.PhaseGate` instead. Then, we can have the phase shift gate `qml.PhaseShift(phi)` get decomposed to `qml.RZ(phi)` up to a global phase if the device doesn't support phase shifts.

josh146 commented on pennylane/ops/qubit.py:
```diff
@@ -253,6 +278,11 @@ def _matrix(cls, *params):
 
+    @staticmethod
+    def decomposition(wires):
+        decomp_ops = [RZ(np.pi / 4, wires=wires)]
```
As with the above case, this could also be done without the global phase factor via `qml.PhaseShift(np.pi / 4, wires=wires)`

josh146 commented on pennylane/ops/qubit.py:
```diff
@@ -193,6 +208,11 @@ def _eigvals(cls, *params):
 
+    @staticmethod
+    def decomposition(wires):
+        decomp_ops = [RZ(np.pi, wires=wires)]
```
(or, as with the others, `qml.PhaseShift(np.pi, wires=wires)`)

josh146 commented on pennylane/ops/qubit.py:
```diff
@@ -116,6 +121,11 @@ def diagonalizing_gates(self):
 
+    @staticmethod
+    def decomposition(wires):
+        decomp_ops = [RX(np.pi, wires=wires)]
```
Or, alternatively,
```python
>>> X = qml.S(wires=0).matrix @ qml.RX(np.pi, wires=0).matrix @ qml.S(wires=0).matrix
>>> np.real_if_close(np.round(X, 3))
array([[ 0.,  1.],
       [ 1., 0.]])
```

josh146 commented on pennylane/ops/qubit.py:
```diff
@@ -161,6 +171,11 @@ def diagonalizing_gates(self):
 
+    @staticmethod
+    def decomposition(wires):
+        decomp_ops = [RY(np.pi, wires=wires)]
```
(or S.RY(pi).S)

josh146 commented on tests/ops/test_qubit_ops.py:
```diff
@@ -330,6 +330,76 @@ def test_matrices(self, ops, mat, tol):
+
+        assert res[1].name == "RZ"
+        assert res[1].wires == [0]
+        assert res[0].params[0] == np.pi / 2
```
Really thorough testing @Lucaman99! I would recommend however including numeric checks to ensure that the decomposition is correct (~up to global phase). For example,
```python
    def test_hadamard_decomposition(self, tol):
        """Tests that the decomposition of the PauliX is correct"""
        op = qml.Hadamard(wires=0)
        res = op.decomposition(0)

        decomposed_matrix = np.linalg.multi_dot([i.matrix for i in reversed(res)])
        global_phase = (decomposed_matrix[op.matrix != 0] / op.matrix[op.matrix != 0])[0]

        assert np.allclose(decomposed_matrix, global_phase * op.matrix, atol=tol, rtol=0)
```

Lucaman99 commented:
Hey @josh146 - I made the requested changes, everything is now in terms of PhaseShift gates. I also re-did the tests to add numerical checks.

Lucaman99 commented:
I also added a method that decomposes the PhaseShift into an RZ gate (as you mentioned above). I'm not totally sure if this is what you meant when you said: "we can have the phase shift gate qml.PhaseShift(phi) get decomposed to qml.RZ(phi) up to a global phase if the device doesn't support phase shifts", but I added it anyways (along with a test).

co9olguy commented on .github/CHANGELOG.md:
```diff
@@ -79,6 +79,9 @@
 <h3>Improvements</h3>
 
+* Adds `decomposition` method to PauliX, PauliY, PauliZ, S, T, and Hadamard gates, which decomposes
+  each of these gates into a product of rotation gates. [(#668)](https://github.com/XanaduAI/pennylane/pull/668)
```
```suggestion
  each of these gates into rotation gates. [(#668)](https://github.com/XanaduAI/pennylane/pull/668)
```

Lucaman99 commented:
Awesome @co9olguy! It seems as though that Travis isn't working because an unrelated test is failing. Do you know why this would be happening/is it related to the changes I made?

Lucaman99 commented:
@co9olguy Just made the changes. I totally forgot that adding in the phase gates to the decompositions removes the need to account for the global phase in all but one of the tests 😆

co9olguy commented:
> Awesome @co9olguy! It seems as though that Travis isn't working because an unrelated test is failing. Do you know why this would be happening/is it related to the changes I made?

Yes, those failures are due to some changes @mariaschuld has been working on. I recently experienced the same thing. No biggie.

@mariaschuld Would it make sense to update the `__eq__` method for the wires class to return `True` also if `wires.tolist() == [some_wires_list]`?

mariaschuld commented:
> > Awesome @co9olguy! It seems as though that Travis isn't working because an unrelated test is failing. Do you know why this would be happening/is it related to the changes I made?
> 
> Yes, those failures are due to some changes @mariaschuld has been working on. I recently experienced the same thing. No biggie.
> 
> @mariaschuld Would it make sense to update the `__eq__` method for the wires class to return `True` also if `wires.tolist() == [some_wires_list]`?

I had a look, and it seems like you just need to write things like `assert op.wires == qml.wires.Wires([0])` in the assert statement. (The `wires` attribute of operators is now a Wires object, so if you want to check it, just compare it with a Wires object.) Hope that fixes everything?

To answer your question, I'm not sure what you mean, since `wires.tolist() == [some_wires_list]` just compares two lists and should be True. If you meant making `Wires([0, 1, 2]) == [0, 1, 2]` True, I think that would be a very bad idea with lots of unwanted side effects.

mariaschuld commented:
@Lucaman99 feel free to ask me if you have any questions! But you may also find help in other operator tests, the assert statements compare PennyLane Wire objects in lots of places.
